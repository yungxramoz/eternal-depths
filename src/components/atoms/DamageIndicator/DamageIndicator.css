.message {
  font-size: 1.5rem;
  position: absolute;
  z-index: 999;
  top: 48%;
  left: 50%;
  transform: translate(-50%, -50%);
  opacity: 0;
  font-size: larger !important;
  text-align: center;
  opacity: 1;
  animation: fadeOutSlideUp 0.7s forwards;
}

@keyframes fadeOutSlideUp {
  0% {
    opacity: 1;
    transform: translate(-50%, -50%) translateY(0);
  }
  100% {
    opacity: 0;
    transform: translate(-50%, -50%) translateY(-15px);
  }
}

.hit {
  color: #fff !important;
}

.critical {
  color: orange !important;
  font-style: bold;
}

.evaded,
.missed {
  color: lightgrey !important;
}
