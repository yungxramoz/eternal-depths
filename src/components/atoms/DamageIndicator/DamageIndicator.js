import React, { useEffect } from 'react'
import './DamageIndicator.css'

const DamageIndicator = ({ messages, setMessages }) => {
  useEffect(() => {
    messages.forEach((message) => {
      setTimeout(() => {
        setMessages((prev) => prev.filter((m) => m.id !== message.id))
      }, 1000)
    })
  }, [messages, setMessages])

  const messageText = (message) => {
    switch (message.type) {
      case 'hit':
        return `-${message.damage}`
      case 'critical':
        return `CRITICAL! -${message.damage}`
      case 'evaded':
        return 'Evaded'
      case 'missed':
        return 'Missed'
      default:
        return ''
    }
  }

  return (
    <>
      {messages.map((message) => (
        <p key={message.id} className={`message ${message.type}`}>
          {messageText(message)}
        </p>
      ))}
    </>
  )
}

export default DamageIndicator
