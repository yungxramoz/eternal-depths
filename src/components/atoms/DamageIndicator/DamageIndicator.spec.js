import React from 'react'
import { render, act } from '@testing-library/react'
import DamageIndicator from './DamageIndicator'
import ATTACK_RESULT from '../../../constants/attack-result'

describe('DamageIndicator', () => {
  let setMessages
  let messages

  beforeEach(() => {
    setMessages = jest.fn()
    messages = [
      { id: 1, type: ATTACK_RESULT.HIT, damage: 10 },
      { id: 2, type: ATTACK_RESULT.CRITICAL, damage: 20 },
      { id: 3, type: ATTACK_RESULT.EVADED },
      { id: 4, type: ATTACK_RESULT.MISSED },
    ]
  })

  it('renders correctly', () => {
    render(<DamageIndicator messages={messages} setMessages={setMessages} />)
    expect(document.body).toMatchSnapshot('DamageIndicator')
  })

  it('calls setMessages to remove messages after 1 second', async () => {
    jest.useFakeTimers()
    render(<DamageIndicator messages={messages} setMessages={setMessages} />)
    jest.advanceTimersByTime(1000)
    expect(setMessages).toHaveBeenCalledTimes(messages.length)
  })
})
