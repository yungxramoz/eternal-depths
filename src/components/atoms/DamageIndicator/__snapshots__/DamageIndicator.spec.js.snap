// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DamageIndicator renders correctly: DamageIndicator 1`] = `
<body>
  <div>
    <p
      class="message hit"
    >
      -10
    </p>
    <p
      class="message critical"
    >
      CRITICAL! -20
    </p>
    <p
      class="message evaded"
    >
      Evaded
    </p>
    <p
      class="message missed"
    >
      Missed
    </p>
  </div>
</body>
`;
